gene/mutate.go

```diff
@@ -0,0 +1,478 @@
+		if val != dna.A && val != dna.C && val != dna.G && val != dna.T {
+			return answer, errors.New("all alt bases must be A, C, T, or G")
+		}
+	}
+	if genomePos < 0 {
+		return answer, errors.New("genomePos must be positive")
+	}
+	if g.posStrand {
+		if genomePos < g.startPos {
+			return answer, errors.New("input genomePos is not in the gene")
+		}
+		genomeIndexPos = genomePos - g.startPos
+	} else {
+		if genomePos > g.startPos {
+			return answer, errors.New("input genomePos is not in the gene")
+		}
+		genomeIndexPos = (g.startPos - genomePos) - 1 // -1 is so the genomeIndexPos is the base BEFORE the insertion
+		dna.ReverseComplement(alt)
+	}
+
+	if genomeIndexPos > len(g.genomeSeq)-1 {
+		return answer, errors.New("input genomePos is not in the gene")
+	}
+
+	// If no error, then append changeLog
+	g.changeLog = append(g.changeLog, log)
+
+	// Update genome sequence
+	g.genomeSeq = append(g.genomeSeq, alt...)                                     // just to increase cap, will get overwritten next line
+	copy(g.genomeSeq[genomeIndexPos+1+len(alt):], g.genomeSeq[genomeIndexPos+1:]) // shift seq over len(alt) positions
+	for idx, val := range alt {                                                   // copy over inserted bases
+		g.genomeSeq[genomeIndexPos+1+idx] = val
+	}
+
+	// Update CDS starts and ends
+	for idx, val := range g.cdsStarts {
+		if val <= genomeIndexPos && g.cdsEnds[idx] > genomeIndexPos {
+			g.cdsEnds[idx] += len(alt)
+		}
+		if val > genomeIndexPos {
+			g.cdsStarts[idx] += len(alt)
+			g.cdsEnds[idx] += len(alt)
+		}
+	}
+
+	if g.featureArray[genomeIndexPos] >= 0 && g.featureArray[genomeIndexPos+1] >= 0 { // Coding
+		cdnaPos := int(g.featureArray[genomeIndexPos])
+
+		// Update featureArray for coding
+		fillVal := Feature(cdnaPos + 1)
+		tmpSlice := make([]Feature, len(alt))
+		g.featureArray = append(g.featureArray, tmpSlice...)
+		copy(g.featureArray[genomeIndexPos+1+len(alt):], g.featureArray[genomeIndexPos+1:])
+		var i int
+		for i = 0; i < len(alt); i++ { // fill the inserted elements with the appropriate cDNA pos value
+			g.featureArray[genomeIndexPos+1+i] = fillVal
+			fillVal++
+		}
+		for ; g.featureArray[genomeIndexPos+1+i] >= 0; i++ { // update the rest of the current exon
+			g.featureArray[genomeIndexPos+1+i] = fillVal
+			fillVal++
+		}
+		for _, cdsStart := range g.cdsStarts { // update the rest of the CDS exons
+			if cdsStart > genomeIndexPos {
+				for k := 0; g.featureArray[cdsStart+k] >= 0; k++ {
+					g.featureArray[cdsStart+k] = fillVal
+					fillVal++
+				}
+			}
+		}
+
+		//TODO: Effect Prediction WIP
+		//answer.CdnaPos, answer.CdnaOffset, err = GenomicPosToCdna(g, genomePos + 1)
+		//frame := (cdnaPos + 1) % 3
+		//var currCodon dna.Codon
+		//
+		//if frame != 0 { // if insertion disrupts a codon, note the codon being changed
+		//	currCodon, err = CdnaPosToCodon(g, cdnaPos)
+		//	answer.AaRef = []dna.AminoAcid{dna.TranslateCodon(&currCodon)}
+		//}
+
+		// Update cDNA
```

Comment:
Maybe dna.Insert could be used here too?

Reply:
Changed